File: app/src/main/java/com/sample/app/fragment_with_activity_result/FragmentWithActivityResult.java
package com.sample.app.fragment_with_activity_result;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.sample.app.R;

import static android.app.Activity.RESULT_OK;
import static com.sample.app.fragment_with_activity_result.ActivityWithResult.KEY_RESULT;

public class FragmentWithActivityResult extends Fragment {
    private TextView resultView;
    static final int REQUEST_TEXT = 0;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.fragment_with_activity_result, container, false);

        resultView = (TextView) rootView.findViewById(R.id.result);
        bindButton(rootView);

        return rootView;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        Log.d("activityResult", "activityResult");
        if (resultCode == RESULT_OK) {
            switch (requestCode) {
                case REQUEST_TEXT:
                    String result = data.getStringExtra(KEY_RESULT);
                    String resultStr = "result is: " + result;
                    resultView.setText(resultStr);
                    break;
            }
        }
    }

    private void bindButton(View rootView) {
        Button button = (Button) rootView.findViewById(R.id.button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                runActivityforResult();
            }
        });
    }

    private void runActivityforResult() {
        Intent intent = new Intent(getActivity(), ActivityWithResult.class);
        startActivityForResult(intent, REQUEST_TEXT);
    }
}

File: app/src/main/java/com/sample/app/fragment_with_activity_result/FragmentWithActivityResult.kt
package com.sample.app.fragment_with_activity_result

import android.app.Activity.RESULT_OK
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import com.sample.app.R
import com.sample.app.fragment_with_activity_result.ActivityWithResult.Companion.KEY_RESULT

class FragmentWithActivityResult : Fragment() {
    private var resultView: TextView? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val rootView = inflater!!.inflate(R.layout.fragment_with_activity_result, container, false)

        resultView = rootView.findViewById(R.id.result) as TextView
        bindButton(rootView)

        return rootView
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        Log.d("activityResult", "activityResult")
        if (resultCode == RESULT_OK) {
            when (requestCode) {
                REQUEST_TEXT -> {
                    val result = data!!.getStringExtra(KEY_RESULT)
                    val resultStr = "result is: " + result
                    resultView!!.text = resultStr
                }
            }
        }
    }

    private fun bindButton(rootView: View) {
        val button = rootView.findViewById(R.id.button) as Button
        button.setOnClickListener { runActivityforResult() }
    }

    private fun runActivityforResult() {
        val intent = Intent(activity, ActivityWithResult::class.java)
        startActivityForResult(intent, REQUEST_TEXT)
    }

    companion object {
        internal val REQUEST_TEXT = 0
    }
}